3.8/graphs-custom-pregel.md

```diff
@@ -0,0 +1,1535 @@
+Therefore we need to get them to know that value.
+
+```json
+"inDegree": {
+  "accumulatorType": "sum",
+  "valueType": "ints"
+}
+```
+
+The choice for `valueType` is equal compared to "outDegree"
+because of the same reason. But as you can see, the `accumulatorType` is now
+set to `sum`. As our vertices do not know how many incoming edges there are,
+each vertex needs to send a message to all outgoing. In our program, a message
+will be sent to every neighbor. That means a vertex with **n** neighbors,
+will send **n** messages. As in our case, a single message represents a single
+incoming edge, we need to add **"+1"** to our accumulator per each message, and
+therefor set the `accumulatorType` to `sum`.
+
+#### Program
+
+##### initProgram
+
+```js
+initProgram: [
+  "seq",
+
+  ["accum-set!", "outDegree", ["this-outbound-edges-count"]], // Sets our outDegree accumulator with ["this-outbound-edges-count"]
+
+  // Init in degree to 0
+  ["accum-set!", "inDegree", 0],             // Initializes our inDegree (sum) accumulator to 0
+  ["send-to-all-neighbours", "inDegree", 1]  // Sends value: "1" to all neighbors, so their inDegree can be raised next round!
+]
+```
+
+##### updateProgram
+
+```
+updateProgram: ["seq",
+  false]
+}]
+```
+
+As in our case, we do not need an update program. We are just inserting a dummy
+(void) method (will be improved in further state). But currently it is necessary,
+because our update program needs to run once to accumulate the inDegrees values
+that have been sent out in our `initProgram`. Therefore `maxGSS` is set to `2`.
+
+#### Storing the result
+
+To be able to store the result, we either need to define a `resultField`
+(which will just store all accumulators into the given resultField attribute)
+or create a `<program>` which will take care of our store procedure.
+The next code snippet demonstrates how a store program could look like:
+
+```js
+1. "dataAccess": {
+2.   "writeVertex": [
+3.     "attrib-set",
+4.       ["attrib-set", ["dict"], "inDegree", ["accum-ref", "inDegree"]], // 1st parameter of outer attrib-set
+5.       "outDegree",                                                     // 2nd parameter of outer attrib-set
+6.       ["accum-ref", "outDegree"]                                       // 3rd parameter of outer attrib-set
+7.   ]
+8. }
+```
+
+We do want to store the value of inDegree and outDegree in our vertices in a
+specific attribute and want to represent them as a JSON Object. Therefore we are
+combining two `attrib-set` commands to achieve that.
+
+The `attrib-set` syntax: `["attrib-set", dict, key, value]`
+
+Let us take a look at row 4:
+
+```js
+[
+  "attrib-set",
+  ["dict"],                  // creates a dict (which is an empty object type)
+  "inDegree",                // the key we want to store as a string
+  ["accum-ref", "inDegree"]  // the accumulator reference we're reading from, which will be inserted in our new dict
+]
+```
```

Comment:
`````suggestion
```js
[
  "attrib-set",

    // creates a dict (which is an empty object type)
  ["dict"],

    // the key we want to store as a string
  "inDegree",

  // the accumulator reference we're reading from, which will be inserted in our new dict
  ["accum-ref", "inDegree"]
]
```
`````